[simgear] error: field ‘_uniforms’ has incomplete type ‘simgear

```
/home/walter/.cache/yay/simgear/src/simgear-2020.3.13/simgear/scene/viewer/Compositor.hxx:137:34: error: field ‘_uniforms’ has incomplete type ‘simgear::compositor::Compositor::BuiltinUniforms’ {aka ‘std::array<osg::ref_ptr<osg::Uniform>, 14>’}
  137 |     BuiltinUniforms              _uniforms;
      |                                  ^~~~~~~~~
In file included from /usr/include/c++/12.2.1/bits/hashtable_policy.h:34,
                 from /usr/include/c++/12.2.1/bits/hashtable.h:35,
                 from /usr/include/c++/12.2.1/unordered_map:46,
                 from /home/walter/.cache/yay/simgear/src/simgear-2020.3.13/simgear/scene/viewer/Compositor.hxx:20:
```

Getting this for both `flight-gear` and `flight-gear-git`

I have exactly the same problem. The error occurs under arch for both packages as well for me, **flight-gear-git** and  **flight-gear**

`[ 95%] Building CXX object simgear/CMakeFiles/SimGearScene.dir/scene/viewer/Compositor.cxx.o
In Datei, eingebunden von /home/hibert/simgear/src/simgear-2020.3.17/simgear/scene/viewer/Compositor.cxx:17:
/home/hibert/simgear/src/simgear-2020.3.17/simgear/scene/viewer/Compositor.hxx:137:34: Fehler: Feld »_uniforms« hat unvollständigen Typen »simgear::compositor::Compositor::BuiltinUniforms« {aka »std::array<osg::ref_ptr<osg::Uniform>, 14>«}
  137 |     BuiltinUniforms              _uniforms;
      |                                  ^~~~~~~~~
In Datei, eingebunden von /usr/include/c++/12.2.1/bits/hashtable_policy.h:34,
                 von /usr/include/c++/12.2.1/bits/hashtable.h:35,
                 von /usr/include/c++/12.2.1/unordered_map:46,
                 von /home/hibert/simgear/src/simgear-2020.3.17/simgear/scene/viewer/Compositor.hxx:20:
/usr/include/c++/12.2.1/tuple:1595:45: Anmerkung: Deklaration von »simgear::compositor::Compositor::BuiltinUniforms« {aka »struct std::array<osg::ref_ptr<osg::Uniform>, 14>«}
 1595 |   template<typename _Tp, size_t _Nm> struct array;
      |                                             ^~~~~
make[2]: *** [simgear/CMakeFiles/SimGearScene.dir/build.make:1826: simgear/CMakeFiles/SimGearScene.dir/scene/viewer/Compositor.cxx.o] Fehler 1
make[1]: *** [CMakeFiles/Makefile2:887: simgear/CMakeFiles/SimGearScene.dir/all] Fehler 2
make: *** [Makefile:156: all] Fehler 2
`

I'm still new to the Linux world. Can you show me what change you made?



> Seems to be a c++11 vs c++ 17 issue... flightgear and simgear both specify 11 in the `CMakeLists.txt` but seems to only build for me when I change to 17. I wonder how others got this buillt in the first place?

unfortunately I can't get any further. Thanks for the help

--- a/simgear/PKGBUILD
+++ b/simgear/PKGBUILD
@@ -4,11 +4,12 @@
 # Contributor: Sergej Pupykin <pupykin.s+arch@gmail.com>
 # Contributor: William Rea <sillywilly@gmail.com>
 # Contributor: Hans Janssen <hans@janserv.xs4all.nl>
+# Contributor: Edoardo Morandi <morandidodo@gmail.com>
 
 pkgname=simgear
 pkgver=2020.3.17
 _pkgver=${pkgver%.*}
-pkgrel=1
+pkgrel=2
 pkgdesc="A set of open-source libraries designed to be used as building blocks for quickly assembling 3d simulations, games, and visualization applications."
 arch=('x86_64')
 url="http://home.flightgear.org/"
@@ -16,8 +17,19 @@ license=('GPL')
 depends=('glu' 'glut' 'freealut' 'plib' 'openscenegraph')
 makedepends=('boost' 'cmake' 'mesa')
 options=('staticlibs')
-source=("https://downloads.sourceforge.net/project/flightgear/release-${_pkgver}/${pkgname}-${pkgver}.tar.bz2")
-sha256sums=('4c5e9cd0dcff1bd311c2460aa601230bd80abdbaec751b559c1d7147749f377c')
+source=(
+    "https://downloads.sourceforge.net/project/flightgear/release-${_pkgver}/${pkgname}-${pkgver}.tar.bz2"
+    "compositor-pass-missing-array-include.patch"
+)
+sha256sums=(
+    '4c5e9cd0dcff1bd311c2460aa601230bd80abdbaec751b559c1d7147749f377c'
+    '81f817ba0f54952ac5fea32ec717275cdb9804f9801def9ae1a00f019bcf968b'
+)
+
+prepare() {
+    cd "$pkgname-$pkgver"
+    patch --forward --strip=1 --input="${srcdir}/compositor-pass-missing-array-include.patch"
+}
 
 build() {
   rm -rf "$srcdir"/simgear-build

--- a/simgear/compositor-pass-missing-array-include.patch
+++ b/simgear/compositor-pass-missing-array-include.patch
@@ -1,0 +1,11 @@
+diff -ur simgear-2020.3.17.orig/simgear/scene/viewer/Compositor.hxx simgear-2020.3.17/simgear/scene/viewer/Compositor.hxx
+--- simgear-2020.3.17.orig/simgear/scene/viewer/Compositor.hxx	2022-11-20 19:01:30.000000000 +0100
++++ simgear-2020.3.17/simgear/scene/viewer/Compositor.hxx	2023-03-10 18:40:59.219808993 +0100
+@@ -17,6 +17,7 @@
+ #ifndef SG_COMPOSITOR_HXX
+ #define SG_COMPOSITOR_HXX
+ 
++#include <array>
+ #include <unordered_map>
+ #include <vector>
+ 

